velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -0,0 +1,452 @@
+      substraitFieldExpr->mutable_direct_reference()->mutable_struct_field();
+
+  const auto& preNodeColNames = inputType->names();
+  const auto& preNodeColTypes = inputType->children();
+  const auto& preNodeColNums = preNodeColNames.size();
+  auto currentColId = -1;
+
+  VELOX_CHECK_EQ(preNodeColNums, preNodeColTypes.size());
+
+  for (auto i = 0; i < preNodeColNums; i++) {
+    if (preNodeColNames[i] == exprName && preNodeColTypes[i] == exprType) {
+      currentColId = i;
+      break;
+    }
+  }
+
+  if (currentColId == -1) {
+    currentColId = preNodeColNums + 1;
+  }
+  directStruct->set_field(currentColId);
+
+  return *substraitFieldExpr;
+}
+
+const ::substrait::Expression& VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const CallTypedExpr>& callTypeExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression* substraitExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression>(&arena);
+
+  auto callTypeExprInputs = callTypeExpr->inputs();
+  std::string callTypeExprFunName = callTypeExpr->name();
+
+  // The processing is different for different function names.
+  // TODO add support for if Expr and switch Expr
+  if (callTypeExprFunName != exec::kIf &&
+      callTypeExprFunName != exec::kSwitch) {
+    ::substrait::Expression_ScalarFunction* scalaExpr =
+        substraitExpr->mutable_scalar_function();
+
+    // TODO need to change yaml file to register function, now is dummy.
+
+    scalaExpr->set_function_reference(
+        funcConvertor_.registerSubstraitFunction(callTypeExprFunName));
+
+    for (auto& arg : callTypeExprInputs) {
+      scalaExpr->add_args()->MergeFrom(toSubstraitExpr(arena, arg, inputType));
+    }
+
+    scalaExpr->mutable_output_type()->MergeFrom(
+        typeConvertor_.toSubstraitType(arena, callTypeExpr->type()));
+
+  } else {
+    VELOX_NYI("Unsupport CallTypeExpr with FunName '{}'", callTypeExprFunName);
+  }
+
+  return *substraitExpr;
+}
+
+const ::substrait::Expression_Literal&
+VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const ConstantTypedExpr>& constExpr,
+    ::substrait::Expression_Literal_Struct* litValue) {
+  if (constExpr->hasValueVector()) {
+    return toSubstraitLiteral(arena, constExpr->valueVector(), litValue);
+  } else {
+    return toSubstraitLiteral(arena, constExpr->value());
+  }
+}
+
+const ::substrait::Expression_Literal&
+VeloxToSubstraitExprConvertor::toSubstraitLiteral(
+    google::protobuf::Arena& arena,
+    const velox::variant& variantValue) {
+  ::substrait::Expression_Literal* literalExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression_Literal>(
+          &arena);
+  switch (variantValue.kind()) {
+    case velox::TypeKind::DOUBLE: {
+      literalExpr->set_fp64(variantValue.value<TypeKind::DOUBLE>());
+      break;
+    }
+    case velox::TypeKind::VARCHAR: {
```

Comment:
Let's make sure we have tests to cover all branches here.

Reply:
Remove VARCHAR/SMALLINT/TINYINT/REAL/TIMESTAMP since Substrait to Velox haven't support them. Could add them in follow up PRs.